File: Examples/FSharp/UserDataWebSocket/UserDataWebSocket_Example.fs
open System
open System.Net
open System.Net.Http
open System.Threading;
open System.Threading.Tasks
open Microsoft.Extensions.Logging
open Binance.Common
open Binance.Spot
open Binance.Spot.Models
open Newtonsoft.Json.Linq

[<EntryPoint>]
let main argv =
    let loggerFactory = LoggerFactory.Create(fun (builder:ILoggingBuilder) -> 
        builder.AddConsole() |> ignore
    )
    let logger = loggerFactory.CreateLogger()

    let loggingHandler = new BinanceLoggingHandler(logger)
    let httpClient = new HttpClient(loggingHandler)

    let apiKey = "vmPUZE6mv9SD5VNHk4HlWFsOr6aKE2zvsw0MuIgwCIPy6utIco14y7Ju91duEh8A";
    let apiSecret = "NhqPtmdSJYdKjVHjA7PZj4Mge3R5YNiP1e3UZjInClVN65XAbvqqM6A7H5fATj0j";
    
    let userDataStreams = new UserDataStreams(httpClient, apiKey = apiKey, apiSecret = apiSecret)
    let response = userDataStreams.CreateSpotListenKey() |> Async.AwaitTask |> Async.RunSynchronously
    let listenKey = JObject.Parse(response).SelectToken("listenKey") |> string

    let websocket = new UserDataWebSocket(listenKey);

    let onlyOneMessage = new TaskCompletionSource<string>();
    websocket.OnMessageReceived(
        (fun data ->
            onlyOneMessage.SetResult(data)
            Task.CompletedTask)
    , CancellationToken.None);
    websocket.ConnectAsync(CancellationToken.None) |> Async.AwaitTask |> Async.RunSynchronously

    let message = onlyOneMessage.Task |> Async.AwaitTask |> Async.RunSynchronously
    logger.LogInformation(message);
    websocket.DisconnectAsync(CancellationToken.None) |> Async.AwaitTask |> Async.RunSynchronously
    
    0

File: Examples/CSharp/UserDataWebSocket/UserDataWebSocket_Example.cs
namespace Binance.Spot.UserDataWebSocketExamples
{
    using System;
    using System.Net;
    using System.Net.Http;
    using System.Threading;
    using System.Threading.Tasks;
    using Binance.Common;
    using Binance.Spot;
    using Binance.Spot.Models;
    using Microsoft.Extensions.Logging;
    using Newtonsoft.Json;

    public class UserDataWebSocket_Example
    {
        public static async Task Main(string[] args)
        {
            using var loggerFactory = LoggerFactory.Create(builder =>
            {
                builder.AddConsole();
            });
            ILogger logger = loggerFactory.CreateLogger<UserDataWebSocket_Example>();
            HttpMessageHandler loggingHandler = new BinanceLoggingHandler(logger: logger);
            HttpClient httpClient = new HttpClient(handler: loggingHandler);

            var apiKey = "api-key";
            var apiSecret = "api-secret";

            var userDataStreams = new UserDataStreams(httpClient, apiKey: apiKey, apiSecret: apiSecret);
            var response = await userDataStreams.CreateSpotListenKey();
            string listenKey = JsonConvert.DeserializeObject<dynamic>(response).listenKey.ToString();

            var websocket = new UserDataWebSocket(listenKey);

            var onlyOneMessage = new TaskCompletionSource<string>();
            websocket.OnMessageReceived(
                async (data) =>
            {
                onlyOneMessage.SetResult(data);
            }, CancellationToken.None);

            await websocket.ConnectAsync(CancellationToken.None);
            string message = await onlyOneMessage.Task;
            logger.LogInformation(message);
            await websocket.DisconnectAsync(CancellationToken.None);
        }
    }
}